File: schemas/auth-rbac/convex/permissions.ts
// permissions: Granular capabilities using a "resource:action" naming convention.
// Assigned to roles (not directly to users).
// See README.md for full design rationale.

import { defineTable } from "convex/server";
import { v } from "convex/values";

export const permissions = defineTable({
  slug: v.string(), // Structured key: "resource:action" (e.g., "posts:create", "users:delete").
  name: v.string(), // Display name (e.g., "Create Posts").
  description: v.optional(v.string()),
  resourceType: v.optional(v.string()), // Groups permissions by resource (e.g., "posts", "users").
})
  .index("by_slug", ["slug"])
  .index("by_resource_type", ["resourceType"]);

File: schemas/auth-rbac/spacetimedb/permissions.rs
// permissions: Granular capabilities using "resource:action" naming convention.
// Assigned to roles (not directly to users) to keep the RBAC model manageable.
// See README.md for full design rationale.

use spacetimedb::Timestamp;

#[spacetimedb::table(name = permissions, public)]
pub struct Permission {
    #[primary_key]
    pub id: String, // UUID

    #[unique]
    pub slug: String, // Structured key: "resource:action" (e.g., "posts:create", "billing:read").

    pub name: String, // Display name (e.g., "Create Posts").

    pub description: Option<String>,

    // Groups permissions by resource (e.g., "posts", "users", "billing").
    // Useful for building permission UIs: "Posts: ☑ create ☑ read ☐ delete".
    #[index(btree)]
    pub resource_type: Option<String>,

    pub created_at: Timestamp,
}
